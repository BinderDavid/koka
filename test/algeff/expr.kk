// --------------------------------------------------------
// Definition of players
// --------------------------------------------------------

type expr {
  Int( i : int )
  Div( e1 : expr, e2: expr )
}

function show( e : expr ) {
  match(e) {
    Int(i) -> core/show(i)
    Div(e1,e2) -> "(" + e1.show + "/" + e2.show + ")"
  }
}

val expr1 = Int(1)
val expr2 = Div(Div(Int(10),Int(2)),Int(3))
val exprX = Div(Div(Int(10),Int(2)),Int(0))

function eval0( e : expr ) : int {
  match(e) {
    Int(i) -> i
    Div(e1,e2) -> {
      val x = eval0(e1)
      val y = eval0(e2)
      x / y
    }
  }
}

// --------------------------------------------------------
// Exceptions
// --------------------------------------------------------

effect exc {
  raise<b>(s : string) : b
}

alias eval0 = <exc>

val hexc = handler {
  return x -> Right(x)
  raise(s) -> Left(s)
}

function showErr( s : a -> string, x : either<string,a> ) : string {
  match(x) {
    Left(l)  -> "exception: " + l
    Right(r) -> s(r)
  }
}

function show1( x : either<string,int> ) {
  showErr(show,x)
}

function eval1( e : expr ) : eval0 int {
  match(e) {
    Int(i) -> i
    Div(e1,e2) -> {
      val x = eval1(e1)
      val y = eval1(e2)
      if (y==0) then raise("divide by zero") else x / y
    }
  }
}

function test1(e) {
  val r = hexc{ eval1(e) }
  println(show1(r))
}
// --------------------------------------------------------
// State
// --------------------------------------------------------

effect tick {
  xtick() : ()
}

alias eval2 = <tick,exc>

val htick = handler(i) {
  return x -> (x,i)
  xtick()   -> resume((),i+1)
}

function show2( x : (either<string,int>,int) ) : string {
  "ticks: " + x.snd.show + "\n" + show1(x.fst)
}

function eval2( e : expr ) : eval2 int {
  match(e) {
    Int(i) -> i
    Div(e1,e2) -> {
      val x = eval2(e1)
      val y = eval2(e2)
      xtick()
      if (y==0) then raise("divide by zero") else x / y
    }
  }
}

function test2(e) {
  htick({ hexc{ eval2(e) }},0).show2.println
}


// --------------------------------------------------------
// Output
// --------------------------------------------------------

effect out {
  line(e:expr) : ()
}

alias eval3 = <out,tick,exc>

val hout = handler {
  return x -> x
  line(e)  -> {
    println(e.show)
    resume(())
  }
}

function eval3( e : expr ) : eval3 int {
  line(e)
  match(e) {
    Int(i) -> i
    Div(e1,e2) -> {
      val x = eval3(e1)
      val y = eval3(e2)
      xtick()
      if (y==0) then raise("divide by zero") else x / y
    }
  }
}

function test3(e) {
  htick({ hout{ hexc{ eval3(e) }}},0) .show2.println
}
