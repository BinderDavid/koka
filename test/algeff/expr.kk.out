2
ticks: 2
2
((16/2)/3)
(16/2)
16
2
3
ticks: 2
2
.Op-failure : forall<a> (s : string) -> .op-failure
.Op-line : (e : expr) -> .op-line
.Op-tick : .op-tick
.Ops-failure: (failure : .op-failure) -> .ops-exc
.Ops-line : (line : .op-line) -> .ops-out
.Ops-tick : (tick : .op-tick) -> .ops-state
.tag-exc : string
.tag-op-failure: string
.tag-op-line: string
.tag-op-tick: string
.tag-out : string
.tag-state : string
>>= : forall<a,b> (m : stexn<a>, f : (a) -> stexn<b>) -> stexn<b>
Div : (e1 : expr, e2 : expr) -> expr
Int : (i : int) -> expr
div? : (expr : expr) -> bool
eval : (e : expr) -> stexn<int>
eval0 : (e : expr) -> int
eval1 : (e : expr) -> exc int
eval2 : (e : expr) -> <exc,state> int
eval3 : (e : expr) -> <exc,out,state> int
exc : forall<a,e> (action : () -> <exc|e> a) -> e either<string,a>
expr1 : expr
expr2 : expr
exprX : expr
failure : forall<a> (s : string) -> exc a
int? : (expr : expr) -> bool
line : (e : expr) -> out ()
main : () -> console ()
mfail : forall<a,b,c,e> (err : a) -> ((s : b) -> e either<a,c>)
mtick : forall<a,e> () -> ((s : int) -> e either<a,((), int)>)
out : forall<a,e> (action : () -> <out,console|e> a) -> <console|e> a
run : forall<a,b,e> (s : a, m : (a) -> e b) -> e b
show : (r : either<string,(int, int)>) -> total string
show : (e : expr) -> total string
show1 : (x : either<string,int>) -> string
show2 : (x : (either<string,int>, int)) -> string
showErr : forall<a> (s : (a) -> string, x : either<string,a>) -> string
state : forall<a,e> (i : int, action : () -> <state|e> a) -> e (a, int)
test1 : (e : expr) -> console ()
test2 : (e : expr) -> console ()
test3 : (e : expr) -> console ()
tick : () -> state ()
unit : forall<a> (x : a) -> stexn<a>